### PR TITLE
OSSMDOC-102

### PR DESCRIPTION
**Pull request by neal-timpe**
https://issues.redhat.com/browse/OSSMDOC-102

WebAssembly extensions replace Mixer

### CODE REVIEW

**Comment by neal-timpe** [comment]
@dgn can you take a look?

File: service_mesh/v2x/ossm-extensions.adoc [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+|priority
+|The default priority of your Extension. This is a required field.
+
+|module
+|The relative path from the container filesystem's root to your WebAssembly module. This is a required field.
+|===
+
+=== Example
+
+For a complete example that was built using the Rust SDK, take a look at the link:https://github.com/maistra/header-append-filter[header-append-filter]. It is a very simple filter that append a header, `custom-header`, to all responses, with the value depending on its configuration.
```
Comment (dgn) [inline service_mesh/v2x/ossm-extensions.adoc:92]:
```suggestion
For a complete example that was built using the Rust SDK, take a look at the link:https://github.com/maistra/header-append-filter[header-append-filter]. It is a very simple filter that appends a header, `custom-header`, to all responses, with the value depending on its configuration.
```

File: service_mesh/v2x/ossm-extensions.adoc [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+|metadata.namespace
+|The `metadata.namespace` of a `ServiceMeshExtension` source has a special semantic: if it equals the Control Plane Namespace, the Extension will be applied to all workloads in the Service Mesh that match its `workloadSelector`. When deployed to any other Mesh Namespace, it will only be applied to workloads in that same Namespace.
+
+|spec.workloadSelector
+|The `spec.workloadSelector` field has the same semantic as the `spec.selector` field of the link:https://istio.io/v1.6/docs/reference/config/networking/gateway/#Gateway[Istio Gateway resource]. It will match a workload based on its Pod labels. If no `workloadSelector` is specified, the Extension will be applied to all workloads in the namespace.
+
+|spec.config
+|This is a pass-through string field that will be handed over to the Extension, so syntax and semantics are dependant on the Extension you're deploying.
+
+|spec.phase
```
Comment (dgn) [inline service_mesh/v2x/ossm-extensions.adoc:158]: this one's a duplicate and can be removed

**Review by dgn — APPROVED** [review]
looking good overall, just two minor things

File: service_mesh/v2x/ossm-extensions.adoc (hunk 2 above)
Comment (tvieira) [reply service_mesh/v2x/ossm-extensions.adoc:158]: I noticed the field is duplicated but the description is not. Just a heads up so you use the right description.

**Review by tvieira — APPROVED** [review]

**Review by gbaufake — APPROVED** [review]

File: service_mesh/v2x/ossm-extensions.adoc [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+|name
+|The name of your extension. This field is just metadata and currently unused.
+
+|description
+|The description of your extension. This field is just metadata and currently unused.
+
+|version
+|The version of your extension. This field is just metadata and currently unused.
+
+|phase
```
Comment (JStickler) [inline service_mesh/v2x/ossm-extensions.adoc:80]: Is there a list of valid values for Phase?

ETA - Found it later in the document.

File: service_mesh/v2x/ossm-extensions.adoc [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+|The default execution phase of your Extension. This is a required field.
+
+|priority
+|The default priority of your Extension. This is a required field.
+
+|module
+|The relative path from the container filesystem's root to your WebAssembly module. This is a required field.
+|===
+
+=== Example
```
Comment (JStickler) [inline service_mesh/v2x/ossm-extensions.adoc:90]: Example Rust extension   (to be more descriptive)

File: service_mesh/v2x/ossm-extensions.adoc [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+{ProductName} Extensions can be enabled using the `ServiceMeshExtension` resource. The following snippet is an example resource.
+
+[source,yaml]
+----
+apiVersion: maistra.io/v1alpha1
+kind: ServiceMeshExtension
+metadata:
+  name: header-append
+  # Deploy to the Control Plane Namespace to enable globally,
```
Comment (JStickler) [inline service_mesh/v2x/ossm-extensions.adoc:121]: Do we need these comments in the example if they're repeated in the table below?  I think the code comments could be merged into the table.

File: service_mesh/v2x/ossm-extensions.adoc [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+|The `metadata.namespace` of a `ServiceMeshExtension` source has a special semantic: if it equals the Control Plane Namespace, the Extension will be applied to all workloads in the Service Mesh that match its `workloadSelector`. When deployed to any other Mesh Namespace, it will only be applied to workloads in that same Namespace.
+
+|spec.workloadSelector
+|The `spec.workloadSelector` field has the same semantic as the `spec.selector` field of the link:https://istio.io/v1.6/docs/reference/config/networking/gateway/#Gateway[Istio Gateway resource]. It will match a workload based on its Pod labels. If no `workloadSelector` is specified, the Extension will be applied to all workloads in the namespace.
+
+|spec.config
+|This is a pass-through string field that will be handed over to the Extension, so syntax and semantics are dependant on the Extension you're deploying.
+
+|spec.image
+|A container image URI pointing to the image that holds the Extension.
```
Comment (JStickler) [inline service_mesh/v2x/ossm-extensions.adoc:159]: Are there any requirements or limitations on where this image is located/hosted?

File: service_mesh/v2x/ossm-extensions.adoc [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+|spec.phase
+|This field defaults to the value set in the `manifest.yaml` of the Extension, but can be overwritten by the user. The phase determines where in the filter chain the Extension is injected, in relation to existing Istio functionality like Authentication, Authorization and metrics generation. Valid values are: PreAuthN, PostAuthN, PreAuthZ, PostAuthZ, PreStats, PostStats. This field defaults to the value set in the `manifest.yaml` of the Extension, but can be overwritten by the user.
```
Comment (JStickler) [inline service_mesh/v2x/ossm-extensions.adoc:162]: And here's the answer to my question about phases.

File: service_mesh/v2x/ossm-extensions.adoc (hunk 6 above)
Comment (dgn) [reply service_mesh/v2x/ossm-extensions.adoc:159]: Any container registry (like e.g. quay.io) will do

**Review by luksa — APPROVED** [review]

File: service_mesh/v2x/ossm-extensions.adoc [hunk 8]
```diff
@@ -3,45 +3,45 @@
 include::modules/ossm-document-attributes.adoc[]
 :context: ossm-extensions
 
-If you are using protocols not yet supported by the proxy, or tokens that {ProductName} doesn't understand, you'd previously have to make sure they are uniformly implemented in all languages used across your infrastructure. You can use WebAssembly Extensions to add that functionality directly into the {ProductName} proxies, allowing you to move even more common functionality out of your applications, and implement them in a single language that compiles to WebAssembly bytecode.
+You can use WebAssembly extensions to add that functionality directly into the {ProductName} proxies, allowing you to move even more common functionality out of your applications, and implement them in a single language that compiles to WebAssembly bytecode.
```
Comment (dgn) [inline service_mesh/v2x/ossm-extensions.adoc:6]:
```suggestion
You can use WebAssembly extensions to add new features directly into the {ProductName} proxies, allowing you to move even more common functionality out of your applications, and implement them in a single language that compiles to WebAssembly bytecode.
```

**Review by JStickler — APPROVED** [review]

**Review by dgn — APPROVED** [review]

**Review by yxun — APPROVED** [review]

**Comment by JStickler** [comment]
@vikram-redhat , any idea what's wrong with the cherry pick robot?  This was merged a week ago.

**Comment by neal-timpe** [comment]
I added the service mesh label an hour ago and I think it tried to add it to the cherry picks and threw an error.

**Pull request merged**